### PR TITLE
INT-3691 Add ZookeeperMetadataStore

### PR DESCRIPTION
**Pull request by mbogoevici**
- create new core interfaces `ListenableMetadataStore` and `MessageStoreListener`;
- create new module `spring-integration-zookeeper`;
- add `ZookeeperMetadataStore` implementation

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/metadata/MessageStoreListenerAdapter.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package org.springframework.integration.metadata;
+
+/**
+ * Base implementation for a {@link MetadataStoreListener}. Subclasses may override any of the methods.
+ *
+ * @author Marius Bogoevici
+ */
+public abstract class MessageStoreListenerAdapter implements MetadataStoreListener {
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/metadata/MessageStoreListenerAdapter.java]: `s/Message/Metadata/` ?

Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/metadata/MessageStoreListenerAdapter.java]: +1

File: spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/Conversions.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.io.UnsupportedEncodingException;
+
+/**
+ *
+ * Utilities for null-safe conversions.
+ *
+ * @author Marius Bogoevici
+ */
+public class Conversions {
```
Comment (garyrussell) [inline spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/Conversions.java]: This should probably go to `-core`; perhaps with a more intuitive name, or maybe just put the methods in `IntegrationUtils` ?

File: spring-integration-zookeeper/src/test/java/org/springframework/integration/zookeeper/metadata/EventuallyMatcher.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * A matcher that will evaluate another matcher repeatedly until it matches, or fail after some number of attempts.
+ *
+ * @param <U> the type the wrapped matcher operates on
+ *
+ * (Copied from rom org.springframework.xd.test.fixtures.EventuallyMatcher)
+ *
+ * @author Eric Bottard
+ */
+public class EventuallyMatcher<U> extends DiagnosingMatcher<U> {
```
Comment (garyrussell) [inline spring-integration-zookeeper/src/test/java/org/springframework/integration/zookeeper/metadata/EventuallyMatcher.java]: Move to `spring-integration-test`

**Comment by garyrussell** [comment]
I guess we gave this a thorough review over in the extensions repo so I haven't reviewed closely this time.

Aside from the minor issues above, we need a `zookeeper.adoc` - but we can defer that until we add the other zk support elements if you prefer.

File: spring-integration-core/src/main/java/org/springframework/integration/metadata/MessageStoreListenerAdapter.java (hunk 1 above)
Comment (mbogoevici) [inline spring-integration-core/src/main/java/org/springframework/integration/metadata/MessageStoreListenerAdapter.java]: Ouch.

File: spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/Conversions.java (hunk 2 above)
Comment (mbogoevici) [inline spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/Conversions.java]: IntegrationUtils is fine for me.

**Comment by mbogoevici** [comment]
Thanks for the review - I would prefer to add doc at least after we do namespace support (which is a requirement too).

**Comment by garyrussell** [comment]
We don't currently have namespace support for other metadata stores so I don't think this, specifically, needs it.

They are configured as `<bean/>` s.

File: spring-integration-core/src/main/java/org/springframework/integration/metadata/ListenableMetadataStore.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package org.springframework.integration.metadata;
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/metadata/ListenableMetadataStore.java]: Oops - needs copyright - will add during merge.

**Comment by garyrussell** [comment]
Merged as 8782d097cf7465efdb69c204ef2f0142b7b1a7cf

File: spring-integration-core/src/main/java/org/springframework/integration/metadata/ListenableMetadataStore.java (hunk 4 above)
Comment (mbogoevici) [inline spring-integration-core/src/main/java/org/springframework/integration/metadata/ListenableMetadataStore.java]: Argh sorry. Every time I reinstall IntelliJ :(

**Pull request closed without merging**